### DISCUSSION TITLE
Step signal pulse width

### DISCUSSION THREAD

**Discussion by aaronknczn** [kachurovskiy/nanoels#105] [Q&A]
Im really happy with my Nano Els H2 and it works great. However the 200 steps per rotation limit makes for quite a noisy operation especially at lower speeds. I have designed a logic circuit that acts as a pulse multiplier. Meaning for one pulse from the arduino it would send three pulses to the stepper so that I can increase the stepper resolution by three and hopefully smooth out the rotation at low spindle speeds. 

My question is this: 
How wide is the pulse width for the step signal coming from the arduino? Is that an adjustable constant or is it dependant on spindle speeds etc. If the latter is true is there a maximum value for the pulse width? 

Because my circuit acts on the falling edge of the pulse very long pulse widths might lead to latency issues.

Thanks in advance!

**Comment by kachurovskiy** [comment]
Hi Aaron,

> How wide is the pulse width for the step signal coming from the arduino?

For normal movements, currently the pulse is 150 microseconds wide which is a lot (there's no space for extra pulses at max RPM) - but you can control it in https://github.com/kachurovskiy/nanoels/blob/main/h2/h2.ino#L1158 and it's easy to make it as short as 4 microseconds by moving the `DLOW(STEP);` line right before the `DHIGH(STEP);`.

![image](https://user-images.githubusercontent.com/517919/214385807-91b8852b-55c0-41b8-bc06-79524fb391b9.png)

For async movements, currently the pulse is 4 microseconds wide.

![image](https://user-images.githubusercontent.com/517919/214385927-e3698433-03c7-44ea-abb0-644f8d9f2f4a.png)

> Is that an adjustable constant or is it dependant on spindle speeds.

As per above, it's 2 constants, which one is used only depends on the mode (ASY or not).

> Because my circuit acts on the falling edge of the pulse very long pulse widths might lead to latency issues.

What you might want to watch out for is the distance between your pulses. It should be calculated as (distance between NanoEls pulses) / (pulse multiplier factor) to keep the overall pulse frequency smooth.

An alternative simple solution might be to get a good driver that is silent at 200 steps e.g. RTELLIGENT R60. StepperOnline drivers are probably also good. However I appreciate the need for an engineering challenge so let's discuss the other option :)

Some follow-up questions:

1. Do you only hear the noise when the half-nut is unlocked? Engaging the half-nut really helps against vibrations.
2. Have you tried 400 or 800 steps setting on your driver? Was the noise gone or reduced? Were you limited by the smaller RPM?

Cheers!

**Reply by aaronknczn** [reply]
Thank you very much for the detailed reply!

> For normal movements, currently the pulse is 150 microseconds wide which is a lot (there's no space for extra pulses at max RPM) - but you can control it in https://github.com/kachurovskiy/nanoels/blob/main/h2/h2.ino#L1158 and it's easy to make it as short as 4 microseconds by moving the DLOW(STEP); line right before the DHIGH(STEP);.

That sounds like I should be able to make it work then.

> An alternative simple solution might be to get a good driver that is silent at 200 steps e.g. RTELLIGENT R60. StepperOnline drivers are probably also good. However I appreciate the need for an engineering challenge so let's discuss the other option :)


I cant Imagine the driver is to blame here. Im using a Leadshine ES-​DH2306. I think I havent done a very good job explaining my problem.
 Im not experiencing electrical noise rather mechanical, in form of the large 20Nm Stepper motor making a knocking sound because of the relatively harsh acceleration and deceleration over the 1,8 degrees of one step. 
![IMG_20230116_204225](https://user-images.githubusercontent.com/121685266/214389798-7dbeefeb-d60e-457e-a5eb-2376e2678d13.jpg)


The cast Iron housing of my lathe acting as a sort of resonance body. Because the lathe is a bit larger than most of the ones Ive seen most people use in this project there is quite a bit more mass that needs to be stopped as well as accelerated. 
When I turned up the step count the knocking went down considerably and at higher RPMs it isnt an issue either because it doesnt really decelerate, ergo also no knocking.

The knocking only shows up when turning the chuck by hand or for a short amount of time when the lathe is accelerating. Its not a huge issue but if it could be resolved by a not very expensive logic circuit I would like to do so.

**Reply by kachurovskiy** [reply]
Got it. If you could share a short video demonstrating your problem it'd be helpful.

> When I turned up the step count the knocking went down considerably and at higher RPMs it isnt an issue either because it doesnt really decelerate, ergo also no knocking.

Can you keep the step count at 800? That would mean limiting the stepper rpm to 200 but that might be sufficient depending on your workloads.

I do have a proper solution for this problem in the works - switching to ESP32 instead of Arduino for the upcoming 2-axis NanoEls - but it will take a couple months of development (2nd iteration of the PCBs are being produced now) and I might or might not be able to ship the PCB to you depending on your country (if you would be interested to begin with 😀).

![image](https://user-images.githubusercontent.com/517919/214392221-b41fae40-c143-4cac-931f-56565a4b33c5.png)

**Reply by aaronknczn** [reply]
Heres a video describing the problem:

https://youtu.be/LzOYAPL3E3U

Granted its a small issue that in no way shape or form relates to the actual function of the ELS but why let that stop me from sinking an unreasonable amount of effort into solving it XD. 

> I do have a proper solution for this problem in the works - switching to ESP32 instead of Arduino for the upcoming 2-axis NanoEls - but it will take a couple months of development (2nd iteration of the PCBs are being produced now) and I might or might not be able to ship the PCB to you depending on your country (if you would be interested to begin with 😀).


I would be VERY interested in that. Just hit me up when its ready! Shipping shouldnt be an issue as Im living in Germany (as you might be able to hear from the video 😅).
By the way I can not thank you enough for making all this information available to everyone for free. Cutting threads with this thing, especially with the automatic stop, almost feels like cheating.

**Reply by kachurovskiy** [reply]
Got it, thanks a lot!
